### PR TITLE
[Xcodeproj] Add support for automatic project generation

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
This adds a `--watch` option to generate-xcodeproj subcommand. SwiftPM
will do the necessary configuration and exec to the watchman-make tool.
If the tool can't be found on the system, it'll emit help information to
install it.

<rdar://problem/32390792> Add support for automatically maintaining the generated Xcode project

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

File: Sources/Basic/Process.swift [hunk 1]
```diff
@@ -195,7 +195,7 @@ public final class Process: ObjectIdentifierProtocol {
     /// Returns true if the given program is present and executable in search path.
```
Comment (hartbit) [inline Sources/Basic/Process.swift]: We probably want to update to documentation to say a path is returned.

File: Sources/Commands/SwiftPackageTool.swift [hunk 2]
```diff
@@ -314,11 +323,15 @@ public class SwiftPackageTool: SwiftTool<PackageToolOptions> {
                 $0.outputPath = $3?.path
             })
         binder.bind(
-            option: generateXcodeParser.add(
+            generateXcodeParser.add(
                 option: "--legacy-scheme-generator", kind: Bool.self,
                 usage: "Use the legacy scheme generator"),
+            generateXcodeParser.add(
+                option: "--watch", kind: Bool.self,
+                usage: "Watch the filesystem and autogenerate the Xcode project if needed"),
```
Comment (hartbit) [inline Sources/Commands/SwiftPackageTool.swift]: Would it be worth being more specific?

`Watch for changes to the Package manifest to regenerate the Xcode project`

File: Sources/Commands/WatchmanHelper.swift [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import POSIX
+import Xcodeproj
+
+struct WatchmanMissingDiagnostic: DiagnosticData {
+    static let id = DiagnosticID(
+        type: WatchmanMissingDiagnostic.self,
+        name: "org.swift.diags.watchman-missing",
+        description: {
+            $0 <<< "this feature requires 'watchman' to work"
+            $0 <<< "\n\n    installation instructions for 'watchman' are available at https://facebook.github.io/watchman/docs/install.html#buildinstall"
```
Comment (hartbit) [inline Sources/Commands/WatchmanHelper.swift]: Why the double-newlines?

File: Tests/CommandsTests/PackageToolTests.swift [hunk 4]
```diff
@@ -489,6 +490,37 @@ final class PackageToolTests: XCTestCase {
...
+            XCTAssertEqual(try fs.readFileContents(script), """
+                #!/usr/bin/env bash
+
+
+                # Autogenerated by SwiftPM. Do not edit!
+
+
+                set -eu
+
+                swift package generate-xcodeproj
```
Comment (hartbit) [inline Tests/CommandsTests/PackageToolTests.swift]: Do we also want a test that makes such that we pass `--xcconfig-overrides` when necessary?

File: Sources/Commands/WatchmanHelper.swift (hunk 3 above)
Comment (aciidb0mb3r) [reply Sources/Commands/WatchmanHelper.swift]: Looks a bit "pretty".

**Pull request merged**